Hamming similarity

openkappa opened the pull request:
This is a pull request in reference to the fuzzy equals/hamming distance part of #149. The method will return early container by container. I have a _slightly_ faster but _significantly_ more complicated implementation to return early when the threshold is breached at the container level that I could propose if we can agree on the method name and whether this function belongs in the library.

This pull request depends on #154.

lemire commented:
Also looks good to me. I could imagine making it faster by going down to the container level but it would be a *lot* more work and would not be always faster. 

@gssiyankai  want to have a look?

gssiyankai commented:
@openkappa 
Can you resolve conflicts, please ?

gssiyankai commented on src/main/java/org/roaringbitmap/RoaringBitmap.java:
```diff
@@ -1265,6 +1265,51 @@ public boolean equals(Object o) {
+      Container left = highLowContainer.getContainerAtIndex(pos1);
+      Container right = other.highLowContainer.getContainerAtIndex(pos2);
+      if(key1 == key2) {
+        budget -= (left.getCardinality()
```
Implementing XorCardinality at the container level is an improvement that sounds promising to me...

openkappa replied:
Certainly.

gssiyankai commented on src/main/java/org/roaringbitmap/RoaringBitmap.java:
```diff
@@ -1265,6 +1265,51 @@ public boolean equals(Object o) {
+      final short key2 = other.highLowContainer.getKeyAtIndex(pos2);
+      Container left = highLowContainer.getContainerAtIndex(pos1);
+      Container right = other.highLowContainer.getContainerAtIndex(pos2);
+      if(key1 == key2) {
```
What about a branchless algorithm here ? all containers will be traversed anyway, as long as ```budget``` is positive and there is no galloping search to overcome branch misprediction penalty ...

openkappa replied:
I don't know if I fully understand. Are you suggesting that if the keys aren't equal, for the sake of argument, key1 < key2 we should advance pos1 until key1 == key2, and decrement budget by left.cardinality for each container? 

I guess the idea behind was that most bitmaps aren't hamming similar, and it's better to get out as soon as the budget for differences is exhausted, hence a position by position by iteration. getCardinality is very cheap though so it might be a bit faster, if I have interpreted the term "gallop" correctly.

gssiyankai replied:
Galloping search is a search algorithm on a sorted array that works in 2 steps. First, you start from a given index and move forward, doubling the size of the step at each iteration, until the current element is greater than the one you are looking for. Then, you do a binary search between the current and the previous element you have seen.

openkappa commented:
I put each suggestion from the review in a separate commit. Let me know if 1057515 was what you meant or not.

gssiyankai commented on src/main/java/org/roaringbitmap/Container.java:
```diff
@@ -122,6 +122,16 @@ public int andCardinality(Container x) {
+   * @return the cardinality of the symmetric difference of the two containers
+   */
+  public int xorCardinality(Container other) {
+    return getCardinality() + other.getCardinality() - 2 * andCardinality(other);
```
My idea was to implement ```xorCardinality``` for each type of container vs another container... but on second thoughts, given that ```getCardinality```  is cheap... except for RunContainer... I think the improvement won't be very big... Let's keep it like this... one can try to implement and benchmark later...

openkappa replied:
Actually this is how I started off quite some time ago with the XOR cardinality pull request. The code is very complicated consisting of 6 separate container level XOR algorithms, mandating a few code duplications and is probably full of the kind of bugs you find 1 year after release. The venn diagram identity has the benefit of being obviously correct, so long as you can get andCardinality right, without much overhead as you note.

gssiyankai commented on src/main/java/org/roaringbitmap/buffer/ImmutableRoaringBitmap.java:
```diff
@@ -1026,6 +1019,53 @@ public boolean equals(Object o) {
+    int pos2 = 0;
+    int budget = tolerance;
+    while(budget >= 0) {
+      short key1 = this.highLowContainer.getKeyAtIndex(pos1);
```
I was thinking about something like this:
```java
public boolean isHammingSimilar(RoaringBitmap other, int tolerance) {
    final int size1 = highLowContainer.size();
    final int size2 = other.highLowContainer.size();
    int pos1 = 0;
    int pos2 = 0;
    int budget = tolerance;
    while(budget >= 0 && pos1 < size1 && pos2 < size2) {
      final int key1 = Util.toIntUnsigned(this.highLowContainer.getKeyAtIndex(pos1));
      final int key2 = Util.toIntUnsigned(other.highLowContainer.getKeyAtIndex(pos2));
      final Container c1 = highLowContainer.getContainerAtIndex(pos1);
      final Container c2 = other.highLowContainer.getContainerAtIndex(pos2);
      budget -= ((Integer.signum(key1 - key2) >>> 31) + ((Integer.signum(key1 - key2) + 1) & 0x1)) * c1.getCardinality();
      budget -= ((Integer.signum(key2 - key1) >>> 31) + ((Integer.signum(key1 - key2) + 1) & 0x1)) * c2.getCardinality();
      budget += ((Integer.signum(key1 - key2) + 1) & 0x1) * 2 * c1.andCardinality(c2);
      pos1 += (Integer.signum(key1 - key2) >>> 31) + ((Integer.signum(key1 - key2) + 1) & 0x1);
      pos2 += (Integer.signum(key2 - key1) >>> 31) + ((Integer.signum(key1 - key2) + 1) & 0x1);
    }
    while(budget >= 0 && pos1 < size1) {
      Container container = highLowContainer.getContainerAtIndex(pos1++);
      budget -= container.getCardinality();
    }
    while(budget >= 0 && pos2 < size2) {
      Container container = other.highLowContainer.getContainerAtIndex(pos2++);
      budget -= container.getCardinality();
    }
    return budget >= 0;
  }
```
I'm not sure if it will speed up computations... maybe it would be safer to benchmark with real dataset before using my method...

openkappa replied:
I definitely misunderstood what you were getting at in your original comment, and actually prefer the original implementation to 1057515. Your implementation is sophisticated and I wouldn't have thought to implement the logic that way. I would prefer that either you commit 

`while(budget >= 0 && pos1 < size1 && pos2 < size2) {
      final int key1 = Util.toIntUnsigned(this.highLowContainer.getKeyAtIndex(pos1));
      final int key2 = Util.toIntUnsigned(other.highLowContainer.getKeyAtIndex(pos2));
      final Container c1 = highLowContainer.getContainerAtIndex(pos1);
      final Container c2 = other.highLowContainer.getContainerAtIndex(pos2);
      budget -= ((Integer.signum(key1 - key2) >>> 31) + ((Integer.signum(key1 - key2) + 1) & 0x1)) * c1.getCardinality();
      budget -= ((Integer.signum(key2 - key1) >>> 31) + ((Integer.signum(key1 - key2) + 1) & 0x1)) * c2.getCardinality();
      budget += ((Integer.signum(key1 - key2) + 1) & 0x1) * 2 * c1.andCardinality(c2);
      pos1 += (Integer.signum(key1 - key2) >>> 31) + ((Integer.signum(key1 - key2) + 1) & 0x1);
      pos2 += (Integer.signum(key2 - key1) >>> 31) + ((Integer.signum(key1 - key2) + 1) & 0x1);
    }`

or we revert to the original implementation and back out 1057515. 

The question is whether you'd prefer to be slightly faster to verify that a bitmap is _not_ hamming similar, and avoid calling andCardinality whenever possible, at the expense of being branchy, or if you'd prefer to verify that a bitmap _is_ hamming similar as fast as possible, at the expense of being slower to rule bitmaps out by calling andCardinality more often than necessary.  Being selfish, I need the second use case, and your implementation may well speed up my code that I want to utilise this method from. A benchmark would definitely help decide.

gssiyankai replied:
Let's rollback to the first version. We can keep this as a future possible improvement.

openkappa commented:
@gssiyankai I have cleaned up the history somewhat and reverted to the original algorithm. As discussed, performance can be tuned later if it becomes an issue.

gssiyankai commented on src/test/java/org/roaringbitmap/buffer/TestContainer.java:
```diff
@@ -102,27 +105,27 @@ public void inotTest10() {
-        new short[] {0, 2, 4, (short) 65190, (short) 65191, (short) 65192, (short) 65193,
-            (short) 65194, (short) 65195, (short) 65196, (short) 65197, (short) 65198,
-            (short) 65199, (short) 65200}));
+            new short[]{0, 2, 4, (short) 65190, (short) 65191, (short) 65192, (short) 65193,
```
This change is also an accidental reformatting I guess...

gssiyankai replied:
@lemire defined a formatting policy here https://github.com/RoaringBitmap/RoaringBitmap/tree/master/style. It's based on google check style.
HTH

openkappa replied:
Sorry I got into a complete mess on a new machine with no sane defaults set up. I have reverted all unintentional formatting errors - could these last two commits be squashed if the PR is merged?

gssiyankai commented:
Yes, absolutely.

gssiyankai commented:
@openkappa 
are you done ? If so, we did not rollback to the first version that consisted in 3 consecutive loops without any inner loops...

openkappa commented:
As of 095f32c there are no inner loops and we are using the first version of the algorithm. I apologise for the level of churn since "cleaning up" :)

gssiyankai commented:
Good.

lemire commented:
@gssiyankai I guess we need to issue a new version of the library?

gssiyankai commented:
Yes, please.

lemire commented:
Release 0.6.41
